AsyncDisplayKit/ASCollectionView.mm

```diff
@@ -333,16 +349,26 @@ - (void)setAsyncDataSource:(id<ASCollectionViewDataSource>)asyncDataSource
   if (asyncDataSource == nil) {
     _asyncDataSource = nil;
     _proxyDataSource = _isDeallocating ? nil : [[ASCollectionViewProxy alloc] initWithTarget:nil interceptor:self];
-    _asyncDataSourceImplementsConstrainedSizeForNode = NO;
-    _asyncDataSourceImplementsNodeBlockForItemAtIndexPath = NO;
+    
+    _asyncDataSourceFlags.asyncDataSourceConstrainedSizeForNode = NO;
+    _asyncDataSourceFlags.asyncDataSourceNodeBlockForItemAtIndexPath = NO;
+    _asyncDataSourceFlags.asyncDataSourceNumberOfSectionsInCollectionView = NO;
+    _asyncDataSourceFlags.asyncDataSourceCollectionViewLockDataSource = NO;
+    _asyncDataSourceFlags.asyncDataSourceCollectionViewUnlockDataSource = NO;
+    _asyncDataSourceFlags.asyncDataSourceCollectionViewConstrainedSizeForNodeAtIndexPath = NO;
   } else {
     _asyncDataSource = asyncDataSource;
     _proxyDataSource = [[ASCollectionViewProxy alloc] initWithTarget:_asyncDataSource interceptor:self];
-    _asyncDataSourceImplementsConstrainedSizeForNode = [_asyncDataSource respondsToSelector:@selector(collectionView:constrainedSizeForNodeAtIndexPath:)];
-    _asyncDataSourceImplementsNodeBlockForItemAtIndexPath = [_asyncDataSource respondsToSelector:@selector(collectionView:nodeBlockForItemAtIndexPath:)];
+    
+    _asyncDataSourceFlags.asyncDataSourceConstrainedSizeForNode = [_asyncDataSource respondsToSelector:@selector(collectionView:constrainedSizeForNodeAtIndexPath:)];
+    _asyncDataSourceFlags.asyncDataSourceNodeBlockForItemAtIndexPath = [_asyncDataSource respondsToSelector:@selector(collectionView:nodeBlockForItemAtIndexPath:)];
+    _asyncDataSourceFlags.asyncDataSourceNumberOfSectionsInCollectionView = [_asyncDataSource respondsToSelector:@selector(numberOfSectionsInCollectionView:)];
+    _asyncDataSourceFlags.asyncDataSourceCollectionViewLockDataSource = [_asyncDataSource respondsToSelector:@selector(collectionViewLockDataSource:)];
```

Comment:
Yes we actually lock the data source in the ASDataController if we try to access the data source synchronously. Look into accessDataSourceSynchronously:withBlock in ASDataController. This is used in _reloadDataWithAnimationOptions:synchronously:completion